### PR TITLE
cp2k: added support for AMD toolchain

### PR DESCRIPTION
**Pull request by amd-toolchain-support**
CP2K:
- Adding AOCC and AMD FFTW support

ELPA:
- Adding AOCC support
- Linking BLAS libs

LIBXSMM:
- Adding AOCC support

### CODE REVIEW

File: var/spack/repos/builtin/packages/elpa/package.py [hunk 1]
```diff
@@ -113,7 +124,8 @@ def configure_args(self):
             'FC={0}'.format(spec['mpi'].mpifc),
             'CXX={0}'.format(spec['mpi'].mpicxx),
             'LDFLAGS={0}'.format(spec['lapack'].libs.search_flags),
-            'LIBS={0}'.format(spec['lapack'].libs.link_flags),
+            'LIBS={0} {1}'.format(
+                spec['lapack'].libs.link_flags, spec['blas'].libs.link_flags),
```
Comment (haampie) [inline var/spack/repos/builtin/packages/elpa/package.py:128]: does this work when you use e.g. openblas which bundles lapack and blas in a single library?

File: var/spack/repos/builtin/packages/libxsmm/package.py [hunk 2]
```diff
@@ -82,7 +82,14 @@ def build(self, spec, prefix):
         ]
 
         # JIT (AVX and later) makes MNK, M, N, or K spec. superfluous
-#       make_args += ['MNK=1 4 5 6 8 9 13 16 17 22 23 24 26 32']
+        # make_args += ['MNK=1 4 5 6 8 9 13 16 17 22 23 24 26 32']
+
+        if '%aocc' in spec:
+            make_args += ['GNU=1']
+            make_args += ['AVX=2']
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/libxsmm/package.py]: I think this option should be conditional on the target.

File: var/spack/repos/builtin/packages/elpa/package.py (hunk 1 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/elpa/package.py:128]: @haampie , yes it works.

File: var/spack/repos/builtin/packages/libxsmm/package.py (hunk 2 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/libxsmm/package.py]: Agree. Addressed the same.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 3]
```diff
@@ -227,6 +230,7 @@ def edit(self, spec, prefix):
             'nvhpc': ['-fast'],
             'cray': ['-O2'],
             'xl': ['-O3'],
+            'aocc': ['-O3', '-ffast-math', '-Hx,47,0x10000008', ],
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/cp2k/package.py]: What is the:
```
-Hx,47,0x10000008
```
flag for?

**Review by alalazo — COMMENTED** [review]
One more question on a compiler flag

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 3 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py]: "-Hx,47,0x10000008" flag is fix for known issue (precision loss) with AOCC v2.2 and v2.3. it will be resolved in upcoming AOCC releases.

Comment (alalazo) [reply var/spack/repos/builtin/packages/cp2k/package.py]: @amd-toolchain-support Can you please add a comment saying that above the lines where the flag appears?

Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py]: @alalazo , added comments as required.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 4]
```diff
@@ -412,12 +424,13 @@ def edit(self, spec, prefix):
...
 
         if '+elpa' in spec:
+            dso_suffix = 'a' if '%aocc' in spec else 'so'
             elpa = spec['elpa']
             elpa_suffix = '_openmp' if '+openmp' in elpa else ''
             elpa_incdir = elpa.headers.directories[0]
 
             fcflags += ['-I{0}'.format(os.path.join(elpa_incdir, 'modules'))]
-            libs.append(os.path.join(elpa.libs.directories[0],
+            libs.append(os.path.join(elpa.prefix.lib,
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cp2k/package.py]: We recommend use of Spack's `join_path` instead of `os.path.join`.  I'm seeing only one use of the former in this package.  Consistency would be nice. :)

Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py]: there were close to 20 occasions where `os.path.join` is used in the package. addressed all of them.

**Review by tldahlgren — DISMISSED** [review]
At this point, from the checks, it looks like you just need to fix a few flake8 errors:

- var/spack/repos/builtin/packages/cp2k/package.py:313: [E127] continuation line over-indented for visual indent
17
- var/spack/repos/builtin/packages/cp2k/package.py:413: [E127] continuation line over-indented for visual indent
18
- var/spack/repos/builtin/packages/cp2k/package.py:415: [E127] continuation line over-indented for visual indent
19
- var/spack/repos/builtin/packages/cp2k/package.py:434: [E127] continuation line over-indented for visual indent

**Comment by amd-toolchain-support** [comment]
> At this point, from the checks, it looks like you just need to fix a few flake8 errors:
> 
> * var/spack/repos/builtin/packages/cp2k/package.py:313: [E127] continuation line over-indented for visual indent
>   17
> * var/spack/repos/builtin/packages/cp2k/package.py:413: [E127] continuation line over-indented for visual indent
>   18
> * var/spack/repos/builtin/packages/cp2k/package.py:415: [E127] continuation line over-indented for visual indent
>   19
> * var/spack/repos/builtin/packages/cp2k/package.py:434: [E127] continuation line over-indented for visual indent

Fixed all the issues

**Review by alalazo — CHANGES_REQUESTED** [review]
Just a minor comment on `dso_suffix`. Everything else LGTM.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 5]
```diff
@@ -394,31 +406,31 @@ def edit(self, spec, prefix):
...
                 ),
-                os.path.join(
+                join_path(
                     spec['metis'].libs.directories[0],
                     'libmetis.{0}'.format(dso_suffix)
                 ),
             ])
 
         if '+elpa' in spec:
+            dso_suffix = 'a' if '%aocc' in spec else 'so'
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/cp2k/package.py]: I think this would be wrong on `platform=darwin`. Possibly:
```suggestion
            dso_suffix = 'a' if '%aocc' in spec else dso_suffix
```
would restore the old behavior for other compilers. Looking at the homepage it seems that `cp2k` may be used on Mac, though @dev-zero can comment more about that if needed.

As a minor comment, having a variable named `dso_suffix` containing the suffix for a static library may be confusing. This is to say that, if it is difficult to change the name of the variable you can maybe add a comment on why having static libraries is requested by `%aocc`, similarly to what was done for the `-Hx, ...` option above.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 6]
```diff
@@ -227,6 +232,7 @@ def edit(self, spec, prefix):
             'nvhpc': ['-fast'],
             'cray': ['-O2'],
             'xl': ['-O3'],
+            'aocc': ['-O3', '-ffast-math', '-Hx,47,0x10000008', ],
```
Comment (dev-zero) [inline var/spack/repos/builtin/packages/cp2k/package.py]: Please don't use `-ffast-math`. Been there. Done that.
And especially not if it means adding some undocumented internal undocumented flag to remedy whatever the optimization is doing.

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 5 above)
Comment (dev-zero) [reply var/spack/repos/builtin/packages/cp2k/package.py]: yes, CP2K is used on macOS, but I doubt someone is going to use AOCC there.

What bothers me is the redefinition of the `dso_suffix` itself: should we decide to rearrange the blocks of code (`if ... in spec:`) or need `dso_suffix` later on, we might suddenly link other stuff statically.
And also: why?

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 6 above)
Comment (haampie) [reply var/spack/repos/builtin/packages/cp2k/package.py:235]: If the tests pass... in some cases you need ffast-math to get auto-vectorization.

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 5 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py]: Removed overwriting condition for `dso_suffix` and implemented it in separate `if .... in spec:`

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 6 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py]: @dev-zero 
- “Hx…..” is a bug-fix required if -O2 or -O3 is used with AOCC 2.2 & 2.3. we are working this bug, you may not see this flag in upcoming AOCC releases.
- by dropping `-ffast-math` we are losing performance

Comment (dev-zero) [reply var/spack/repos/builtin/packages/cp2k/package.py]: So, will the flag be still there and work in those future releases of AOCC or should this be guarded by a version check on the compiler from the start?
About performance, I guess it increases performance, but by how much, and on what?
Please note that the `tests` in CP2K are regression tests and will not catch numerical errors which can accumulate over long running simulations.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 7]
```diff
@@ -263,6 +269,12 @@ def edit(self, spec, prefix):
                 '-ffree-line-length-none',
                 '-ggdb',  # make sure we get proper Fortran backtraces
             ]
+        elif '%aocc' in spec:
+            fcflags += [
+                '-D__F2008',
```
Comment (dev-zero) [inline var/spack/repos/builtin/packages/cp2k/package.py]: is this preprocessor flag used by the compiler?

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 8]
```diff
@@ -263,6 +269,12 @@ def edit(self, spec, prefix):
...
+                '-ffree-form',
+            ]
+            ldflags += ['-Wl,-z,muldefs']
```
Comment (dev-zero) [inline var/spack/repos/builtin/packages/cp2k/package.py]: Why is this needed? According to documentation this requires that libraries for linking are passed in a specific order. Is this automatically ensured? If yes, how? If not: which libraries are affected?

Comment (dev-zero) [reply var/spack/repos/builtin/packages/cp2k/package.py]: The reason why I am insisting on this is that with libraries like COSMA, we already have an overloading of `p*gemm`, depending on how it is linked and it is hard to figure out which implementation is actually called when a user reports a performance regression. Hence we have to be able to reason about which implementation is called and why.
Also, static linking can make this even more complicated.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 9]
```diff
@@ -416,11 +427,19 @@ def edit(self, spec, prefix):
...
 
-            fcflags += ['-I{0}'.format(os.path.join(elpa_incdir, 'modules'))]
-            libs.append(os.path.join(elpa.libs.directories[0],
-                                     ('libelpa{elpa_suffix}.{dso_suffix}'
-                                      .format(elpa_suffix=elpa_suffix,
-                                              dso_suffix=dso_suffix))))
+            fcflags += ['-I{0}'.format(join_path(elpa_incdir, 'modules'))]
+
+            # As a best practice used by CP2K developers,
+            # AOCC considering using static libraries of ELPA
```
Comment (dev-zero) [inline var/spack/repos/builtin/packages/cp2k/package.py]: Then please drop it. The only [reason why CP2K is linking to ELPA statically](https://github.com/cp2k/cp2k/commit/6cf6e4d1c367a5a1f5f27c4844edd8583f652c9f#diff-8de432b66f5b5350591f4adab306dc691ededc81d56268b2a5786094668fb313) in the toolchain scripts was the difficulty of ensuring `-fPIC` and managing the different sets of OpenMP-enabled/-disabled libraries. Both should not be an issue with Spack.

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 6 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py]: Removed `-fast-math` and `-Hx` .
`-O3 -ffast-math -Hx` would be 6% better performant over `O1`. Tests with H2O-dft-ls show Energy results diverge at 9th decimal place.

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 7 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py]: Not necessarily required for the compiler, removing  `-D__F2008`.

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 8 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py]: No performance regression observed without `ldflags += ['-Wl,-z,muldefs']` hence removing the same.

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 9 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py]: Currently AOCC supports static libraries of ELPA and support for shared library is ongoing.

**Comment by alalazo** [comment]
@dev-zero Can you please check if all your concerns have been addressed?

@amd-toolchain-support Can you let the reviewer / maintainer resolve conversations that are not trivial? In that way it would be clearer if they think the issue raised has been addressed or not.

**Comment by amd-toolchain-support** [comment]
> @dev-zero Can you please check if all your concerns have been addressed?
> 
> @amd-toolchain-support Can you let the reviewer / maintainer resolve conversations that are not trivial? In that way it would be clearer if they think the issue raised has been addressed or not.

Sure @alalazo .
@dev-zero , can you please review the latest code and provide your comments.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 10]
```diff
@@ -192,18 +192,21 @@ def makefile(self):
...
+        return [join_path(self.stage.source_path, self.makefile)]
 
     def edit(self, spec, prefix):
         pkgconf = which('pkg-config')
 
         if '^fftw' in spec:
             fftw = spec['fftw:openmp' if '+openmp' in spec else 'fftw']
             fftw_header_dir = fftw.headers.directories[0]
+        elif '^amdfftw' in spec:
+            fftw = spec['amdfftw:openmp' if '+openmp' in spec else 'amdfftw']
```
Comment (dev-zero) [inline var/spack/repos/builtin/packages/cp2k/package.py:209]: this probably requires a `conflicts('^amdfftw~openmp', when='+openmp')`(similar to fftw) to avoid silent/strange errors should the user not have `amdfftw+openmp` right from the start

**Comment by dev-zero** [comment]
other than the comment about adding a `conflict` it looks good to me

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 10 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/cp2k/package.py:209]: Agree, addressed the same.

Comment (dev-zero) [reply var/spack/repos/builtin/packages/cp2k/package.py:209]: thanks!

**Review by dev-zero — APPROVED** [review]

**Review by alalazo — APPROVED** [review]
